### PR TITLE
Add generate-themes-doc.fish

### PR DESCRIPTION
**Pull request by Pyppe**
Wanted to browse thru all the available themes, but could not find a centralized place for this. This PR contains a `bin/generate-themes-doc.fish` which generates `docs/Themes.md` that can be used for previewing the themes.

See https://github.com/Pyppe/oh-my-fish/blob/master/docs/Themes.md

### CODE REVIEW

**Comment by bpinto** [comment]
I love this!!! Is there a way to automate it though?

**Comment by derekstavis** [comment]
Thanks!! I like having this docs!! By the way, the script could be improved by generating a table of contents on the top of the docs. What do you think about?

**Comment by Pyppe** [comment]
Thanks for the feedback, guys! I improved the script, and added TOC.

Regarding automation; I don't have a silver bullet. Of course a custom script could be added that overrides the generated `docs/Themes.md` and pushes it to master if it was changed... And then this could be invoked e.g. for each Travis CI-build. But personally I wouldn't want this kind of automagic behavior for my project... :)

Hmm... after-thought: Travis could generate the `Themes.md` and send email if it has changed. (Although, I don't know how easy/possible this kind of automation is to implement in Travis).

**Comment by Pyppe** [comment]
Regarding automation, I think this could be one solution https://github.com/Pyppe/oh-my-fish/commit/a898f84691dcf133c0783d33fb1ec0a55db6a3b7: It runs the `generate-themes-doc.fish` as a Travis-test, and the build fails if `Themes.md` has changed.

Travis seems to be having some issues (https://twitter.com/pyrylahti/status/645578947560873985); so I'm yet to verify that this actually works...

**Comment by derekstavis** [comment]
Looks awesome with the TOC!! I was also wondering if we should include a "Back to top" link after every theme, as navigation gets hard on such big files.

**Comment by derekstavis** [comment]
Regarding to the place where the script is stored, maybe we should add this category of scripts to a `tools` directory, @oh-my-fish/core?

**Comment by bpinto** [comment]
I like using Travis for the automation! Very clever. Maybe tools is a good folder.

**Comment by derekstavis** [comment]
So I propose the following:

    bin/generate-themes-docs.fish -> tools/generate-themes-docs.fish

As I believe that PR #52 will also produce some other scripts that fit into `tools` directory.

**Comment by Pyppe** [comment]
* Changed the directory from `bin/` to `tools/`
* Updated `README.md` to mention `Themes.md`

IMO we could already merge this as is. I can add the Travis integration later once #52 gets merged (because it seems to do some of the same things I'd like to do).

File: tools/generate-themes-doc.fish [hunk 1]
```diff
@@ -0,0 +1,55 @@
+#!/usr/bin/fish
+
+set -l project_dir (status -f|xargs dirname|xargs dirname)
+set -l theme_doc "$project_dir/docs/Themes.md"
+set temp_theme_contents (mktemp /tmp/fish.Themes.Content.XXXXX.md)
```
Comment (bpinto) [inline tools/generate-themes-doc.fish]: Does mktemp work on Linux and OS X?

File: tools/generate-themes-doc.fish [hunk 2]
```diff
@@ -0,0 +1,55 @@
+#!/usr/bin/fish
```
Comment (bpinto) [inline tools/generate-themes-doc.fish]: Get fish binary using env?

File: tools/generate-themes-doc.fish [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+
+set -l project_dir (status -f|xargs dirname|xargs dirname)
+set -l theme_doc "$project_dir/docs/Themes.md"
+set temp_theme_contents (mktemp /tmp/fish.Themes.Content.XXXXX.md)
+set temp_theme_toc (mktemp /tmp/fish.Themes.TOC.XXXXX.md)
+
+function write_theme_readme -a name raw_content readme
```
Comment (bpinto) [inline tools/generate-themes-doc.fish]: Can we follow the guidelines for private functions?

File: tools/generate-themes-doc.fish [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+      echo $readme
+      break
+    end
+  end
+end
+
+echo "Generating Themes documentation ..."
+echo "# Available themes" > $temp_theme_toc
+
+for t in (command find $project_dir/db/themes/ -type f|sort)
```
Comment (bpinto) [inline tools/generate-themes-doc.fish]: Theme instead of t?

File: tools/generate-themes-doc.fish [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+  set -l raw_content (echo $url|sed -r 's#https://github.com/([-.a-z0-9]+)/([-.a-z0-9]+)#https://raw.githubusercontent.com/\1/\2/master#gi')
+  set -l readme (find_readme $raw_content)
+
+  echo "- [$name](#$name)" >> $temp_theme_toc
+  echo "# $name" >> $temp_theme_contents
+  if [ $readme ]
+    echo "Fetching readme for $name"
+    write_theme_readme $name $raw_content $readme
+  else
+    echo "FAILED: No readme for $name"
```
Comment (bpinto) [inline tools/generate-themes-doc.fish]: Output to stderr?

File: tools/generate-themes-doc.fish (hunk 1 above)
Comment (Pyppe) [inline tools/generate-themes-doc.fish]: IMO, yes.

**Comment by Pyppe** [comment]
Made changes. Some questions, though:
* What do you mean by "Get fish binary using env?"
* Renamed scripts functions with `__` prefix? Is this enough / OK?

**Comment by bpinto** [comment]
The prefix is perfect! Regarding env, this is what I mean: http://stackoverflow.com/questions/2429511/why-do-people-write-usr-bin-env-python-on-the-first-line-of-a-python-script

**Comment by Pyppe** [comment]
Roger. Changed the shebang-line to `#!/usr/bin/env fish`

**Comment by bpinto** [comment]
Really nice!!! Eager to see the automation using Travis and perhaps a similar feature for plugins?

:+1: for the merge. Would you mind squashing your commits into one?

**Comment by derekstavis** [comment]
I've tested here, no issues. :+1: for the merge. Also, rebasing into current master would be great to keep the history linear :)

**Comment by Pyppe** [comment]
Rebased and squashed. :shipit:

**Comment by derekstavis** [comment]
Super!! Thanks @Pyppe!!

**Comment by bpinto** [comment]
@Pyppe we are back. Whenever you have a chance to work on the travis script, submit a PR! :dancers:

**Comment by Pyppe** [comment]
I've got the functionality here: https://github.com/Pyppe/oh-my-fish/commit/de464e56b1b401b5e9bc83f604ffe16b4d991244

However, I think we'd need some changes from the #52 PR in order to run these fish-based scripts within Travis. See https://travis-ci.org/Pyppe/oh-my-fish/builds/82388150

**Comment by bpinto** [comment]
It has been merged! :dancers:

**Pull request merged**